src/NServiceBus.Persistence.CosmosDB/Saga/SagaPersister.cs

```diff
@@ -1,87 +1,86 @@
+            var partitionKey = sagaData.Id.ToString();
+            var sagaType = context.GetSagaType();
+            var sagaDataType = sagaData.GetType();
+            var jObject = JObject.FromObject(sagaData);
 
-        public Task Update(IContainSagaData sagaData, SynchronizedStorageSession session, ContextBag context)
-        {
-            return InvokeWrapInDocumentAsGenericMethod(sagaData, context, (document, sagaId, partitionKey) =>
+            jObject.Add("PersisterVersion", FileVersionRetriever.GetFileVersion(typeof(SagaPersister)));
+            jObject.Add("SagaType", sagaType.FullName);
+            jObject.Add("SagaTypeVersion", FileVersionRetriever.GetFileVersion(sagaType));
+            jObject.Add("SagaDataType", sagaDataType.FullName);
+            jObject.Add("SagaDataTypeVersion", FileVersionRetriever.GetFileVersion(sagaDataType));
+
+            using (var stream = new MemoryStream())
+            using (var streamWriter = new StreamWriter(stream))
+            using (JsonWriter jsonWriter = new JsonTextWriter(streamWriter))
             {
-                // only delete if we have the same version as in CosmosDB
-                context.TryGet<string>("cosmosdb_etag", out var etag);
-                var options = new ItemRequestOptions { IfMatchEtag = etag };
+                await jObject.WriteToAsync(jsonWriter).ConfigureAwait(false);
 
-                return container.ReplaceItemAsync(document, sagaId, partitionKey, options);
-            });
+                await container.CreateItemStreamAsync(stream, new PartitionKey(partitionKey)).ConfigureAwait(false);
+            }
         }
 
-        static async Task InvokeWrapInDocumentAsGenericMethod(IContainSagaData sagaData, ContextBag context, Func<object, string, PartitionKey?, Task<ItemResponse<object>>> operation)
+        public Task Update(IContainSagaData sagaData, SynchronizedStorageSession session, ContextBag context)
         {
-            // Save and Update methods require generic method to be executed - build a generic method
-            var sagaDataType = sagaData.GetType();
-            var genericMethod = wrapInDocumentMethod.MakeGenericMethod(sagaDataType);
-
             var partitionKey = sagaData.Id.ToString();
-            var document = genericMethod.Invoke(null, new object[] { sagaData, partitionKey, context });
-            var response = await operation(document, sagaData.Id.ToString(), new PartitionKey(partitionKey)).ConfigureAwait(false);
 
-            context.Set("cosmosdb_etag", response.ETag);
-        }
-
-        static CosmosDbSagaDocument<TSagaData> WrapInDocument<TSagaData>(IContainSagaData sagaData, string partitionKey, ContextBag context) where TSagaData : IContainSagaData
-        {
-            var sagaType = context.GetSagaType();
-            var sagaDataType = sagaData.GetType();
+            // only update if we have the same version as in CosmosDB
+            context.TryGet<string>("cosmosdb_etag", out var etag);
```

Comment:
what happens if there is no such an etag? Can `ItemRequestsOptions` deal if a `null` etag? I will still merge this